pkg/hubble/metrics/metric_config_watcher.go

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package metrics
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+var metricReloadInterval = 10 * time.Second
+
+type metricConfigWatcher struct {
+	logger         logrus.FieldLogger
+	configFilePath string
+	callback       func(ctx context.Context, hash uint64, config api.Config)
+	ticker         *time.Ticker
+	stop           chan bool
+	currentCfgHash uint64
+	cfgStore       map[string]*api.MetricConfig
+	mutex          lock.RWMutex
+}
+
+// NewmetricConfigWatcher creates a config watcher instance. Config watcher notifies
+// DynamicFlowProcessor when config file changes and dynamic metric config should be
+// reconciled.
+func NewMetricConfigWatcher(
+	configFilePath string,
+	callback func(ctx context.Context, hash uint64, config api.Config),
+) *metricConfigWatcher {
+	watcher := &metricConfigWatcher{
+		logger:         logrus.New().WithField(logfields.LogSubsys, "hubble").WithField("configFilePath", configFilePath),
+		configFilePath: configFilePath,
+		callback:       callback,
+		currentCfgHash: 0,
+		cfgStore:       make(map[string]*api.MetricConfig),
+	}
+
+	// initial configuration load
+	watcher.reload()
+
+	watcher.ticker = time.NewTicker(metricReloadInterval)
+	watcher.stop = make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-watcher.stop:
+				return
+			case <-watcher.ticker.C:
+				watcher.reload()
+			}
+		}
+	}()
+
+	return watcher
+}
+
+func (c *metricConfigWatcher) reload() {
+	c.logger.Debug("Attempting reload")
+	config, isSameHash, hash, err := c.readConfig()
+	if err != nil {
+		c.logger.WithError(err).Error("failed reading dynamic exporter config")
+	} else {
+		if !isSameHash {
+			c.callback(context.TODO(), hash, *config)
+		}
+	}
+}
+
+func (c *metricConfigWatcher) resetCfgPath(path string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.configFilePath = path
+}
+
+// Stop stops watcher.
+func (c *metricConfigWatcher) Stop() {
+	if c.ticker != nil {
+		c.ticker.Stop()
+	}
+	c.stop <- true
+}
+
+func (c *metricConfigWatcher) readConfig() (*api.Config, bool, uint64, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	config := &api.Config{Metrics: []*api.MetricConfig{}}
+	yamlFile, err := os.ReadFile(c.configFilePath)
```

Comment:
One thing I'm wondering about, in most cases where Cilium is checking for changes to a file we use the `pkg/fswatcher` package instead of polling. Could that make sense in this instance?

Reply:
I think it could, perhaps as a future improvement. For now, I just followed the example of the dynamic exporter.
Compared to the exporter, this code also bails earlier when it detects the file hash has not changed, rather than unmarshalling the yaml first.

Moreover, I noticed problems when trying this API: text editors will issue multiple update events before a file is saved. So unless we can guarantee the configMap update is always done atomically by the user, I'd rather keep it as is for now. How does that sound?